components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1830,19 +1841,101 @@ where
+            match ctx
+                .engine
+                .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
+            {
+                Ok(None) => (),
+                Ok(Some(state)) => {
+                    *reason = Some(format!("state {:?} exist in kv engine", state));
+                }
+                e => panic!(
+                    "{} failed to get regions state of {}: {:?}",
+                    self.tag, region_id, e
+                ),
+            }
+        }
+
+        // Note that the following execution sequence is possible.
+        // Apply thread:    check `RegionLocalState`(None)
+        // Store thread:    create peer
+        // Peer thread:     apply snapshot and then be destroyed
+        // Apply thread:    check `StoreMeta` and find it's ok to create this new region.
+        // It's **very unlikely** to happen because the step 2 and step 3 should take far more time than the time interval
+        // between step 1 and step 4. (A similiar case can happen in create-peer process, see details in `maybe_create_peer`)
+        // Even it happens, this new region will be destroyed in future when it communicates to other TiKVs or PD.
+        // Now it seems there is no other side effects.
+        let mut meta = ctx.store_meta.lock().unwrap();
+        for (region_id, (peer_id, reason)) in new_regions_map.iter_mut() {
+            if reason.is_some() {
+                continue;
+            }
+            if let Some(r) = meta.regions.get(region_id) {
+                if util::is_region_initialized(r) {
+                    *reason = Some(format!("region {:?} has already initialized", r));
+                } else {
+                    // If the region in `meta.regions` is not initialized, it must exist in `meta.pending_create_peers`.
+                    let status = meta.pending_create_peers.get_mut(region_id).unwrap();
+                    // If they are the same peer, the new one from splitting can replace it.
+                    // Because it must be uninitialized. See detailes in `check_snapshot`.
+                    if *status == (*peer_id, false) {
+                        *status = (*peer_id, true);
+                    } else {
+                        *reason = Some(format!("status {:?} is not expected", status));
+                    }
+                }
+            } else {
+                assert_eq!(
+                    meta.pending_create_peers
+                        .insert(*region_id, (*peer_id, true)),
+                    None
+                );
+            }
+        }
+        drop(meta);
+
+        let kv_wb_mut = ctx.kv_wb.as_mut().unwrap();
+        for new_region in &regions {
+            if new_region.get_id() == derived.get_id() {
+                continue;
+            }
+            let (new_peer_id, reason) = new_regions_map.get(&new_region.get_id()).unwrap();
+            if let Some(r) = reason {
+                warn!(
```

Comment:
Should be info.

Reply:
In normal cases, it's very unlikely to happen. If it really happens, it means this TiKV runs too slow. On the other hand, it may happen if the id allocated from PD turns back. So I think using `warn` is better.